File: tools/elf_image.py
"""Host validation of ELF inputs to the seL4-compatible boot handoff."""
import struct

PAGE = 4096
KERNEL_OFFSET = 0xffff000000000000
KERNEL_LINK_BASE = 0xffff800000000000


def require(condition, message):
    if not condition:
        raise ValueError(message)


def parse_elf(data):
    require(len(data) >= 64, 'truncated ELF header')
    require(data[:7] == b'\x7fELF\x02\x01\x01', 'expected little-endian ELF64')
    header = struct.unpack_from('<HHIQQQIHHHHHH', data, 16)
    kind, machine, version, entry, phoff = header[:5]
    ehsize, phsize, phnum = header[7:10]
    require((kind, machine, version, ehsize, phsize) == (2, 183, 1, 64, 56), 'unsupported ELF header')
    require(0 < phnum <= 32 and phoff + phnum * phsize <= len(data), 'invalid program headers')
    segments, pages = [], set()
    for index in range(phnum):
        typ, flags, offset, va, pa, filesz, memsz, align = struct.unpack_from('<IIQQQQQQ', data, phoff + index * phsize)
        require(typ not in (2, 3, 7), 'dynamic loading/interpreter/TLS unsupported')
        if typ != 1 or memsz == 0:
            continue
        require(flags in (4, 5, 6), 'unsupported segment permissions')
        require(filesz <= memsz and offset + filesz <= len(data), 'truncated segment')
        require(va + memsz < 2**64 and pa + memsz < 2**64, 'segment address overflow')
        require(va % PAGE == 0 and offset % PAGE == 0, 'unaligned segment')
        require(align >= PAGE and align & (align - 1) == 0 and va % align == offset % align, 'invalid alignment')
        end = (va + memsz + PAGE - 1) // PAGE * PAGE
        require(end - va <= 32 * 1024 * 1024, 'segment exceeds platform limit')
        mapped = set(range(va, end, PAGE))
        require(not pages & mapped, 'segments share pages')
        pages |= mapped
        segments.append(dict(va=va, pa=pa, end=end, flags=flags, filesz=filesz, offset=offset, memsz=memsz))
    require(entry % 4 == 0 and any(s['flags'] == 5 and s['va'] <= entry < s['va'] + s['filesz'] for s in segments), 'invalid executable entry')
    return dict(entry=entry, start=min(pages), end=max(pages) + PAGE, segments=segments)


def root_layout(root, dtb_size):
    """Mirror the initial-task resource policy; ELF owns all image/stack VAs."""
    require(PAGE <= root['start'] < root['end'] <= 128 * 1024 * 1024,
            'root outside user address space')
    require(root['start'] % PAGE == root['end'] % PAGE == 0, 'unaligned root image')
    require(root['end'] - root['start'] <= 1024 * PAGE, 'root image span exceeds limit')
    require(40 <= dtb_size <= 1024 * 1024, 'invalid DTB size')
    ipc = root['end']
    boot_info = ipc + PAGE
    extra = boot_info + PAGE
    extra_size = dtb_size + 16
    end = extra + (extra_size + PAGE - 1) // PAGE * PAGE
    require(end <= 128 * 1024 * 1024, 'root metadata exceeds user address space')
    image_pages = sum((s['end'] - s['va']) // PAGE for s in root['segments'])
    require(image_pages + (end - ipc) // PAGE <= 1024, 'root mapped pages exceed limit')
    return dict(ipc=ipc, boot_info=boot_info, extra=extra, extra_size=extra_size, end=end)


def validate_pair(kernel, root, dtb_size, kernel_physical=0x40200000):
    # Input preflight, not the loader's allocation decision. PT_LOAD.p_paddr
    # does not constrain placement; kernel VA is mapped to the chosen RAM gap.
    require(KERNEL_LINK_BASE <= kernel['start'] < kernel['end'] <= KERNEL_LINK_BASE + 0x2000000,
            'kernel outside virtual image window')
    require(kernel['start'] % 0x200000 == 0, 'kernel virtual start must be block aligned')
    root_layout(root, dtb_size)
    dtb_start = kernel_physical + kernel['end'] - kernel['start']
    image_start = (dtb_start + dtb_size + PAGE - 1) // PAGE * PAGE
    image_end = image_start + root['end'] - root['start']
    require(0x40200000 <= kernel_physical and image_end + PAGE <= 0x48000000, 'loaded images exceed RAM')
    return image_start, image_end

File: projects/libs/user/src/elf_image.rs
//! Borrowed ELF64 program headers; no segment arrays on the small user stack.
const PAGE: usize = 4096;
#[derive(Clone, Copy)]
pub struct Segment {
    pub flags: usize,
    pub offset: usize,
    pub va: usize,
    pub filesz: usize,
    pub memsz: usize,
    pub end: usize,
}
pub struct Elf<'a> {
    headers: &'a [u8],
    pub start: usize,
    pub end: usize,
    pub entry: usize,
}
fn integer(data: &[u8], offset: usize, size: usize) -> Result<usize, ()> {
    let data = data
        .get(offset..offset.checked_add(size).ok_or(())?)
        .ok_or(())?;
    Ok(data
        .iter()
        .enumerate()
        .fold(0, |n, (i, byte)| n | ((*byte as usize) << (i * 8))))
}
fn segment(header: &[u8]) -> Result<Segment, ()> {
    let va = integer(header, 16, 8)?;
    let memsz = integer(header, 40, 8)?;
    Ok(Segment {
        flags: integer(header, 4, 4)?,
        offset: integer(header, 8, 8)?,
        va,
        filesz: integer(header, 32, 8)?,
        memsz,
        end: va
            .checked_add(memsz)
            .and_then(|n| n.checked_add(PAGE - 1))
            .ok_or(())?
            & !(PAGE - 1),
    })
}
impl<'a> Elf<'a> {
    pub fn parse(bytes: &'a [u8]) -> Result<Self, ()> {
        if bytes.get(..7) != Some(b"\x7fELF\x02\x01\x01")
            || integer(bytes, 16, 2)? != 2
            || integer(bytes, 18, 2)? != 183
            || integer(bytes, 20, 4)? != 1
            || integer(bytes, 52, 2)? != 64
            || integer(bytes, 54, 2)? != 56
        {
            return Err(());
        }
        let count = integer(bytes, 56, 2)?;
        if !(1..=32).contains(&count) {
            return Err(());
        }
        let phoff = integer(bytes, 32, 8)?;
        let headers = bytes
            .get(phoff..phoff.checked_add(count * 56).ok_or(())?)
            .ok_or(())?;
        let mut elf = Self {
            headers,
            start: usize::MAX,
            end: 0,
            entry: integer(bytes, 24, 8)?,
        };
        let mut entry_valid = false;
        for (index, header) in headers.as_chunks::<56>().0.iter().enumerate() {
            let kind = integer(header, 0, 4)?;
            if matches!(kind, 2 | 3 | 7) {
                return Err(());
            }
            if kind != 1 || integer(header, 40, 8)? == 0 {
                continue;
            }
            let s = segment(header)?;
            let align = integer(header, 48, 8)?;
            if !matches!(s.flags, 4..=6)
                || s.filesz > s.memsz
                || s.offset
                    .checked_add(s.filesz)
                    .is_none_or(|end| end > bytes.len())
                || !s.va.is_multiple_of(PAGE)
                || !s.offset.is_multiple_of(PAGE)
                || align < PAGE
                || !align.is_power_of_two()
                || s.va % align != s.offset % align
                || s.va < 0x400000
                || s.end > 0x8000000
            {
                return Err(());
            }
            for previous in headers[..index * 56].as_chunks::<56>().0.iter() {
                if integer(previous, 0, 4)? == 1 && integer(previous, 40, 8)? != 0 {
                    let other = segment(previous)?;
                    if s.va < other.end && other.va < s.end {
                        return Err(());
                    }
                }
            }
            elf.start = elf.start.min(s.va);
            elf.end = elf.end.max(s.end);
            entry_valid |= s.flags == 5 && s.va <= elf.entry && elf.entry < s.va + s.filesz;
        }
        if !entry_valid || !elf.entry.is_multiple_of(4) {
            return Err(());
        }
        Ok(elf)
    }
    pub fn segments(&self) -> impl Iterator<Item = Segment> + '_ {
        self.headers
            .as_chunks::<56>()
            .0
            .iter()
            .filter(|h| integer(*h, 0, 4) == Ok(1) && integer(*h, 40, 8) != Ok(0))
            .map(|h| segment(h).expect("validated ELF header"))
    }
}
